src/Conversion/ONNXToTOSA/Math/Conv2D.cpp

```diff
@@ -0,0 +1,180 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//===---------------- Conv2D.cpp - Conv2D Op -----------------------------===//
+//
+// Copyright (c) 2022 Advanced Micro Devices, Inc.
+//
+// =============================================================================
+//
+// This file lowers ONNX conv operator to TOSA dialect.
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Dialect/Tosa/IR/TosaOps.h"
+#include "src/Conversion/ONNXToTOSA/DialectBuilder.hpp"
+#include "src/Conversion/ONNXToTOSA/ONNXToTOSACommon.hpp"
+#include "src/Conversion/ONNXToTOSA/ONNXToTOSALegalizeUtils.hpp"
+#include "src/Dialect/ONNX/ONNXOps/ShapeHelper.hpp"
+#include <src/Dialect/Mlir/IndexExpr.hpp>
+
+#include "mlir/Dialect/Tosa/IR/TosaOps.h"
+#include "src/Conversion/ONNXToTOSA/DialectBuilder.hpp"
+#include "src/Conversion/ONNXToTOSA/ONNXToTOSACommon.hpp"
+#include "src/Conversion/ONNXToTOSA/ONNXToTOSALegalizeUtils.hpp"
+#include "src/Dialect/ONNX/ONNXOps/ShapeHelper.hpp"
+#include <src/Dialect/Mlir/IndexExpr.hpp>
+
+using namespace mlir;
+
+namespace onnx_mlir {
+
+namespace {
+
+/// When setting the groups parameters, we have to create multiple conv2d ops
+/// where the input, kernel and bias is a slice of the original inputs.
+/// Afterwards we have to concat the results into a single tensor
+Value createConvInGroups(PatternRewriter &rewriter, Operation *op,
+    TosaBuilder &tosaBuilder, Type &resultType,
+    const llvm::ArrayRef<int64_t> weightShape, Value &newInput,
+    Value &newWeight, Value &bias, const int64_t groups,
+    DenseI64ArrayAttr &pads, DenseI64ArrayAttr &strides,
+    DenseI64ArrayAttr &dilations) {
+  // Set up constants outside of loop
+  const int64_t sizeOfSliceInput = weightShape[1];
+  const int64_t sizeOfSliceKernel = weightShape[0] / groups;
```

Comment:
I'm pretty sure the definition requires that `weightShape[0] % groups == 0` but wonder if you want to assert is with a comment that legally this must hold? I'm not certain if it is needed - I personally just get paranoid reading code with integer `/` without the `%` being used/checked.

Reply:
The ONNX verifier actually checks for that and will not even read the IR if this is incorrect